benchmarks/htap/lib/controller.py

```diff
@@ -16,98 +16,82 @@
 
 from s64da_benchmark_toolkit.dbconn import DBConn
 
-
-def oltp_worker(
-        worker_id, dsn, scale_factor, dry_run, monitoring_interval, reporting_queue, date_value
-):
-    with DBConn(dsn) as conn:
-        start_date = datetime.fromtimestamp(date_value.value)
-        tpcc_worker = Transactions(worker_id, scale_factor, start_date, conn)
-        reporting_interval = timedelta(seconds=monitoring_interval) / 2
-        last_reporting_time = None
-        while True:
-            tpcc_worker.next_transaction(dry_run)
-            now = datetime.now()
-            if last_reporting_time is None or (now - last_reporting_time) >= reporting_interval:
-                reporting_queue.put_nowait(('tpcc', tpcc_worker.stats(worker_id)))
-                last_reporting_time = now
-
-
-def olap_worker(worker_id, dsn, olap_timeout_seconds, dry_run, reporting_queue, date_value, min_date):
-    with DBConn(dsn, statement_timeout=olap_timeout_seconds * 1000) as conn:
-        queries = Queries(worker_id, conn)
-        while True:
-            max_date = datetime.fromtimestamp(date_value.value)
-            queries.run_next_query(min_date, max_date, reporting_queue, dry_run)
-
-
 class HTAPController:
+    # have the shared-memory primitives static as otherwise the multiprocessing
+    # inheritance scheme doesn't work. we want these primitives so we can use
+    # "simple" synchronized primitives.
+    latest_timestamp = Value('d', 0) # database record timestamp
+    next_tsx_timestamp = Value('d', 0) # rtc time at which we can do the next tpcc tsx
+    stats_queue = Queue() # queue for communicating statistics
+
     def __init__(self, args):
-        self.dsn = args.dsn
+        self.args = args
+        self.next_tsx_timestamp.value = datetime.now().timestamp()
+        self.tsx_timestamp_increment = 1.0 / self.args.target_tps if self.args.target_tps is not None else 0
```

Comment:
this can be an argument error and not a division by 0 error

Reply:
if someone deliberately puts in `--target-tps=0` then they can get an error ;)